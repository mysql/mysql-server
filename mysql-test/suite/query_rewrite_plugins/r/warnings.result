#
# Test of rules where the parser gives warnings about the originial
# query, pattern or replacement.
#
call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");
# Query rewrite plugin was installed.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT * FROM query_rewrite.rewrite_rules;
id	pattern	pattern_database	replacement	enabled	message
1	LOAD DATA INFILE "tab1.data" INTO TABLE test.t1	NULL	LOAD DATA INFILE "tab2.data" INTO TABLE test.t1	YES	Statement type of pattern not supported.
2	DO CONCAT("a", "b")	NULL	DO CONCAT("c", "d")	YES	Statement type of pattern not supported.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
# We want the note about the query being rewritten here ...
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
SHOW ERRORS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
@p1
45000
SELECT @p2;
@p2
Loading of some rule(s) failed.
# ... but not here, since we failed parsing after seeing SHOW WARNINGS.
SHOW WARNINGS and give me a parse error, please;
ERROR 42000: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'and give me a parse error, please' at line 1
DELETE FROM query_rewrite.rewrite_rules;
# Test of warnings when we rewrite the diagnostics statement.  echo #
# We disable mtr's SHOW WARNINGS here, as it would just be confusing
# when it gets rewritten without us seeing it.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
( 'SELECT 1', 'SHOW WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT 1;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
Warnings:
Warning	1620	Plugin is busy and will be uninstalled on shutdown
# Query rewrite plugin was queued for uninstalling.
