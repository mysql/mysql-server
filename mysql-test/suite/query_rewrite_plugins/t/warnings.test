--echo #
--echo # Test of rules where the parser gives warnings about the originial
--echo # query, pattern or replacement.
--echo #

call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");

--source suite/query_rewrite_plugins/include/have_plugin_rewriter.inc
--source suite/query_rewrite_plugins/include/install_rewriter.inc

# Still many unsupported statements
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
         'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
       ( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();
SELECT * FROM query_rewrite.rewrite_rules;

INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

--echo # We want the note about the query being rewritten here ...
SHOW WARNINGS;
SHOW ERRORS;
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
SELECT @p2;

--echo # ... but not here, since we failed parsing after seeing SHOW WARNINGS.
--error ER_PARSE_ERROR
SHOW WARNINGS and give me a parse error, please;

DELETE FROM query_rewrite.rewrite_rules;

--echo # Test of warnings when we rewrite the diagnostics statement.  echo #
--echo # We disable mtr's SHOW WARNINGS here, as it would just be confusing
--echo # when it gets rewritten without us seeing it.
--disable_warnings
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
       ( 'SELECT 1', 'SHOW WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

SELECT 1;

SHOW WARNINGS;

--enable_warnings
--source suite/query_rewrite_plugins/include/uninstall_rewriter.inc
