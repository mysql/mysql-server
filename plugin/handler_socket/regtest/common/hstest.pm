
# vim:sw=2:ai

package hstest;

use DBI;
use Net::HandlerSocket;

our %conf = ();

sub get_conf_env {
  my ($key, $defval) = @_;
  return $ENV{$key} || $defval;
}

sub init_conf {
  $conf{host} = get_conf_env("MYHOST", "localhost");
  $conf{myport} = get_conf_env("MYPORT", 3306);
  $conf{dbname} = get_conf_env("MYDBNAME", "hstestdb");
  $conf{ssps} = get_conf_env("MYSSPS");
  $conf{user} = get_conf_env("MYSQLUSER", "root");
  $conf{pass} = get_conf_env("MYSQLPASS", "");
  $conf{hsport} = get_conf_env("HSPORT", 9998);
  $conf{hspass} = get_conf_env("HSPASS", undef);
}

sub get_dbi_connection {
  my ($dbname, $host, $myport, $ssps, $user, $pass)
    = ($conf{dbname}, $conf{host}, $conf{myport}, $conf{ssps},
      $conf{user}, $conf{pass});
  my $mycnf = "binary_my.cnf";
  my $dsn = "DBI:mysql:database=;host=$host;port=$myport"
    . ";mysql_server_prepare=$ssps"
    . ";mysql_read_default_group=perl"
    . ";mysql_read_default_file=../common/$mycnf";
  my $dbh = DBI->connect($dsn, $user, $pass, { RaiseError => 1 });
  return $dbh;
}

sub init_testdb {
  my $charset = $_[0] || "binary";
  my $dbh = get_dbi_connection();
  my $dbname = $conf{dbname};
  $dbh->do("drop database if exists $dbname");
  $dbh->do("create database $dbname default character set $charset");
  $dbh->do("use $dbname");
  return $dbh;
}

sub get_hs_connection {
  my ($host, $port) = @_;
  $host ||= $conf{host};
  $port ||= $conf{hsport};
  my $hsargs = { 'host' => $host, 'port' => $port };
  my $conn = new Net::HandlerSocket($hsargs);
  if (defined($conn) && defined($conf{hspass})) {
    $conn->auth($conf{hspass});
  }
  return $conn;
}


init_conf();

1;

