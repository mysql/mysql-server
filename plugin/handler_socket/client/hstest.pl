#!/usr/bin/perl

# vim:sw=8:ai:ts=8

use strict;
use warnings;

use DBI;
use Net::HandlerSocket;

my %conf = ();
for my $i (@ARGV) {
	my ($k, $v) = split(/=/, $i);
	$conf{$k} = $v;
}

my $verbose = get_conf("verbose", 0);
my $actions_str = get_conf("actions", "hsread");
my $tablesize = get_conf("tablesize", 10000);
my $db = get_conf("db", "hstest");
my $table = get_conf("table", "hstest_table1");
my $engine = get_conf("engine", "innodb");
my $host = get_conf("host", "localhost");
my $mysqlport = get_conf("mysqlport", 3306);
my $mysqluser = get_conf("mysqluser", "root");
my $mysqlpass = get_conf("mysqlpass", "");
my $hsport = get_conf("hsport", 9999);
my $loop = get_conf("loop", 10000);
my $op = get_conf("op", "=");
my $ssps = get_conf("ssps", 0);
my $num_moreflds = get_conf("moreflds", 0);
my $moreflds_prefix = get_conf("moreflds_prefix", "column0123456789_");
my $keytype = get_conf("keytype", "varchar(32)");
my $file = get_conf("file", undef);

my $dsn = "DBI:mysql:database=;host=$host;port=$mysqlport"
	. ";mysql_server_prepare=$ssps";
my $dbh = DBI->connect($dsn, $mysqluser, $mysqlpass, { RaiseError => 1 });
my $hsargs = { 'host' => $host, 'port' => $hsport };
my $cli = new Net::HandlerSocket($hsargs);

my @actions = split(/,/, $actions_str);
for my $action (@actions) {
	if ($action eq "table") {
		print("TABLE $db.$table\n");
		$dbh->do("drop database if exists $db");
		$dbh->do("create database $db");
		$dbh->do("use $db");
		my $moreflds = get_createtbl_moreflds_str();
		$dbh->do(
			"create table $table (" .
			"k $keytype primary key" .
			",v varchar(32) not null" .
			$moreflds .
			") character set utf8 collate utf8_bin " .
			"engine = $engine");
	} elsif ($action eq "insert") {
		print("INSERT $db.$table tablesize=$tablesize\n");
		$dbh->do("use $db");
		my $moreflds = get_insert_moreflds_str();
		for (my $i = 0; $i < $tablesize; $i += 100) {
			my $qstr = "insert into $db.$table values";
			for (my $j = 0; $j < 100; ++$j) {
				if ($j != 0) {
					$qstr .= ",";
				}
				my $k = "" . ($i + $j);
				my $v = "v" . int(rand(1000)) . ($i + $j);
				$qstr .= "('$k', '$v'";
				for (my $j = 0; $j < $num_moreflds; ++$j) {
					$qstr .= ",'$j'";
				}
				$qstr .= ")";
			}
			$dbh->do($qstr);
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "read") {
		print("READ $db.$table op=$op loop=$loop\n");
		$dbh->do("use $db");
		my $moreflds = get_select_moreflds_str();
		my $sth = $dbh->prepare(
			"select k,v$moreflds from $db.$table where k = ?");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			$sth->execute($k);
			if ($verbose >= 10) {
				print "RET:";
				while (my $ref = $sth->fetchrow_arrayref()) {
					my $rk = $ref->[0];
					my $rv = $ref->[1];
					print " $rk $rv";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsinsert") {
		print("HSINSERT $db.$table tablesize=$tablesize\n");
		$cli->open_index(1, $db, $table, '', 'k,v');
		for (my $i = 0; $i < $tablesize; ++$i) {
			my $k = "" . $i;
			my $v = "v" . int(rand(1000)) . $i;
			my $r = $cli->execute_insert(1, [ $k, $v ]);
			if ($r->[0] != 0) {
				die;
			}
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsread") {
		print("HSREAD $db.$table op=$op loop=$loop\n");
		my $moreflds = get_select_moreflds_str();
		$cli->open_index(1, $db, $table, '', "k,v$moreflds");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			my $r = $cli->execute_find(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				my $len = scalar(@{$r});
				print "LEN=$len";
				for my $e (@{$r}) {
					print " [$e]";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsupdate") {
		my $vbase = "v" . int(rand(1000));
		print("HSUPDATE $db.$table op=$op loop=$loop vbase=$vbase\n");
		$cli->open_index(1, $db, $table, '', 'v');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			my $v = $vbase . $i;
			print "k=$k v=$v\n";
			my $r = $cli->execute_update(1, $op, [ $k ], 1, 0,
				[ $v ]);
			if ($verbose >= 10) {
				print "UP k=$k v=$v\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsdelete") {
		print("HSDELETE $db.$table op=$op loop=$loop\n");
		$cli->open_index(1, $db, $table, '', '');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			print "k=$k\n";
			my $r = $cli->execute_delete(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				print "DEL k=$k\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "verify") {
		verify_do();
	}
}

sub verify_do {
	my ($fail_cnt, $ok_cnt) = (0, 0);
	my $sth = $dbh->prepare("select v from $db.$table where k = ?");
	use FileHandle;
	my $fh = new FileHandle($file, "r");
	while (my $line = <$fh>) {
		chomp($line);
		my @vec = split(/\t/, $line);
		my $k = $vec[3];
		my $v = $vec[7];
		next if (!defined($k) || !defined($v));
		# print "$k $v\n";
		$sth->execute($k);
		my $aref = $sth->fetchrow_arrayref();
		if (!defined($aref)) {
			print "FAILED: $k notfound\n";
			++$fail_cnt;
		} else {
			my $gv = $aref->[0];
			if ($gv ne $v) {
				print "FAILED: $k got=$gv expected=$v\n";
				++$fail_cnt;
			} else {
				print "OK: $k $v $gv\n" if $verbose >= 10;
				++$ok_cnt;
			}
		}
	}
	print "OK=$ok_cnt FAIL=$fail_cnt\n";
}

sub get_conf {
	my ($key, $def) = @_;
	my $val = $conf{$key};
	if ($val) {
		print "$key=$val\n";
	} else {
		$val = $def;
		$def ||= '';
		print "$key=$def(default)\n";
	}
	return $val;
}

sub get_createtbl_moreflds_str {
	my $s = "";
	for (my $j = 0; $j < $num_moreflds; ++$j) {
		$s .= ",$moreflds_prefix$j varchar(30)";
	}
	return $s;
}

sub get_select_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",$moreflds_prefix$i";
	}
	return $s;
}

sub get_insert_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",?";
	}
	return $s;
}

