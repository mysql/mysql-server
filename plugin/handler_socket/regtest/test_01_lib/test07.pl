#!/usr/bin/perl

# vim:sw=2:ai

# test for nulls

BEGIN {
	push @INC, "../common/";
};

# use strict;
use warnings;
use hstest;

my $dbh = hstest::init_testdb();
my $table = 'hstesttbl';
my $tablesize = 100;
$dbh->do(
  "create table $table (" .
  "k int primary key, v1 varchar(30), v2 varchar(30), " .
  "key idxv1 (v1) " .
  ") engine = innodb");
srand(999);

my %valmap = ();

my $sth = $dbh->prepare("insert into $table values (?,?,?)");
for (my $i = 0; $i < $tablesize; ++$i) {
  my $k = "" . $i;
  my $v1 = "1v" . int(rand(1000)) . $i;
  my $v2 = "2v" . int(rand(1000)) . $i;
  if ($i % 10 == 3) {
    $v1 = undef;
  }
  $sth->execute($k, $v1, $v2);
  $valmap{$k} = $v1;
}

print "MY\n";
my $aref = $dbh->selectall_arrayref("select k,v1,v2 from $table order by k");
for my $row (@$aref) {
  my ($k, $v1, $v2) = @$row;
  $v1 = "[NULL]" if (!defined($v1));
  print "$k $v1 $v2\n";
}

print "HS\n";
my $hs = hstest::get_hs_connection();
my $dbname = $hstest::conf{dbname};
$hs->open_index(1, $dbname, $table, '', 'k,v1,v2');
my $r = $hs->execute_single(1, '>=', [ '' ], 10000, 0);
shift(@$r);
for (my $i = 0; $i < $tablesize; ++$i) {
  my $k = $r->[$i * 3];
  my $v1 = $r->[$i * 3 + 1];
  my $v2 = $r->[$i * 3 + 2];
  $v1 = "[NULL]" if (!defined($v1));
  print "$k $v1 $v2\n";
}

print "2ndIDX\n";
$hs->open_index(2, $dbname, $table, 'idxv1', 'k,v1,v2');
for (my $i = 0; $i < $tablesize; ++$i) {
  my $k = "" . $i;
  my $v1 = $valmap{$k};
  next if !defined($v1);
  my $r = $hs->execute_single(2, '=', [ $v1 ], 1, 0);
  shift(@$r);
  my $r_k = $r->[0];
  my $r_v1 = $r->[1];
  my $r_v2 = $r->[2];
  print "2ndidx $k $v1 => $r_k $r_v1 $r_v2\n";
}

print "2ndIDX NULL\n";
{
  my %rvals = ();
  my $v1 = undef;
  my @arr;
  push(@arr, undef);
  my $kv = \@arr;
  my $r = $hs->execute_single(2, "=", $kv, 10000, 0);
  shift(@$r);
  for (my $i = 0; $i < scalar(@$r); $i += 3) {
    my $k = $r->[$i];
    my $v1 = $r->[$i + 1];
    my $v2 = $r->[$i + 2];
    $rvals{$k} = [ $k, $v1, $v2 ];
  }
  for my $i (sort { $a <=> $b } keys %rvals) {
    my $rec = $rvals{$i};
    my $k = $rec->[0];
    my $v1 = $rec->[1];
    my $v2 = $rec->[2];
    print "2ndidxnull $k $v2\n";
  }
}

