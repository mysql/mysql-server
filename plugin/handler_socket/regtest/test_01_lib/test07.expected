MY
0 1v1020 2v6350
1 1v8031 2v9251
2 1v7752 2v5372
3 [NULL] 2v4143
4 1v5904 2v3024
5 1v7045 2v7515
6 1v3676 2v4006
7 1v3977 2v1707
8 1v7198 2v7348
9 1v5879 2v4949
10 1v52310 2v95410
11 1v43311 2v82011
12 1v28312 2v83712
13 [NULL] 2v41513
14 1v54514 2v58314
15 1v5215 2v32315
16 1v61416 2v67916
17 1v80517 2v45117
18 1v11518 2v26918
19 1v21819 2v61719
20 1v87820 2v34520
21 1v51221 2v96921
22 1v40822 2v29122
23 [NULL] 2v95323
24 1v71024 2v14224
25 1v68225 2v93425
26 1v62126 2v96526
27 1v57427 2v20427
28 1v29828 2v13428
29 1v98329 2v44429
30 1v14430 2v15230
31 1v18731 2v21531
32 1v832 2v69732
33 [NULL] 2v28033
34 1v70134 2v53734
35 1v41335 2v6935
36 1v8636 2v82236
37 1v67037 2v37037
38 1v80638 2v68838
39 1v2639 2v6639
40 1v80240 2v17140
41 1v55741 2v84741
42 1v77742 2v73042
43 [NULL] 2v11543
44 1v64644 2v49644
45 1v12045 2v68445
46 1v37446 2v6546
47 1v37047 2v17447
48 1v82848 2v86748
49 1v75949 2v70349
50 1v84350 2v94250
51 1v40151 2v36251
52 1v36752 2v30752
53 [NULL] 2v16753
54 1v79954 2v82954
55 1v53955 2v37955
56 1v56056 2v85856
57 1v57957 2v2657
58 1v88758 2v50758
59 1v34559 2v89859
60 1v43060 2v80160
61 1v84561 2v32561
62 1v62462 2v48062
63 [NULL] 2v67663
64 1v5364 2v73664
65 1v80965 2v27065
66 1v17566 2v83966
67 1v6167 2v22267
68 1v35868 2v50568
69 1v62769 2v90669
70 1v4670 2v98170
71 1v11971 2v471
72 1v1472 2v48072
73 [NULL] 2v40573
74 1v87574 2v63974
75 1v82775 2v34475
76 1v59876 2v56376
77 1v77077 2v51677
78 1v53878 2v54878
79 1v35779 2v32279
80 1v3680 2v37080
81 1v33181 2v81581
82 1v76982 2v66882
83 [NULL] 2v28183
84 1v69684 2v45284
85 1v40685 2v185
86 1v39586 2v32486
87 1v3687 2v73887
88 1v16088 2v7988
89 1v75989 2v65789
90 1v3190 2v78390
91 1v65091 2v82491
92 1v85292 2v6892
93 [NULL] 2v54693
94 1v81394 2v77594
95 1v8795 2v69695
96 1v19696 2v38096
97 1v7997 2v75197
98 1v32398 2v21798
99 1v3799 2v70199
HS
0 1v1020 2v6350
1 1v8031 2v9251
2 1v7752 2v5372
3 [NULL] 2v4143
4 1v5904 2v3024
5 1v7045 2v7515
6 1v3676 2v4006
7 1v3977 2v1707
8 1v7198 2v7348
9 1v5879 2v4949
10 1v52310 2v95410
11 1v43311 2v82011
12 1v28312 2v83712
13 [NULL] 2v41513
14 1v54514 2v58314
15 1v5215 2v32315
16 1v61416 2v67916
17 1v80517 2v45117
18 1v11518 2v26918
19 1v21819 2v61719
20 1v87820 2v34520
21 1v51221 2v96921
22 1v40822 2v29122
23 [NULL] 2v95323
24 1v71024 2v14224
25 1v68225 2v93425
26 1v62126 2v96526
27 1v57427 2v20427
28 1v29828 2v13428
29 1v98329 2v44429
30 1v14430 2v15230
31 1v18731 2v21531
32 1v832 2v69732
33 [NULL] 2v28033
34 1v70134 2v53734
35 1v41335 2v6935
36 1v8636 2v82236
37 1v67037 2v37037
38 1v80638 2v68838
39 1v2639 2v6639
40 1v80240 2v17140
41 1v55741 2v84741
42 1v77742 2v73042
43 [NULL] 2v11543
44 1v64644 2v49644
45 1v12045 2v68445
46 1v37446 2v6546
47 1v37047 2v17447
48 1v82848 2v86748
49 1v75949 2v70349
50 1v84350 2v94250
51 1v40151 2v36251
52 1v36752 2v30752
53 [NULL] 2v16753
54 1v79954 2v82954
55 1v53955 2v37955
56 1v56056 2v85856
57 1v57957 2v2657
58 1v88758 2v50758
59 1v34559 2v89859
60 1v43060 2v80160
61 1v84561 2v32561
62 1v62462 2v48062
63 [NULL] 2v67663
64 1v5364 2v73664
65 1v80965 2v27065
66 1v17566 2v83966
67 1v6167 2v22267
68 1v35868 2v50568
69 1v62769 2v90669
70 1v4670 2v98170
71 1v11971 2v471
72 1v1472 2v48072
73 [NULL] 2v40573
74 1v87574 2v63974
75 1v82775 2v34475
76 1v59876 2v56376
77 1v77077 2v51677
78 1v53878 2v54878
79 1v35779 2v32279
80 1v3680 2v37080
81 1v33181 2v81581
82 1v76982 2v66882
83 [NULL] 2v28183
84 1v69684 2v45284
85 1v40685 2v185
86 1v39586 2v32486
87 1v3687 2v73887
88 1v16088 2v7988
89 1v75989 2v65789
90 1v3190 2v78390
91 1v65091 2v82491
92 1v85292 2v6892
93 [NULL] 2v54693
94 1v81394 2v77594
95 1v8795 2v69695
96 1v19696 2v38096
97 1v7997 2v75197
98 1v32398 2v21798
99 1v3799 2v70199
2ndIDX
2ndidx 0 1v1020 => 0 1v1020 2v6350
2ndidx 1 1v8031 => 1 1v8031 2v9251
2ndidx 2 1v7752 => 2 1v7752 2v5372
2ndidx 4 1v5904 => 4 1v5904 2v3024
2ndidx 5 1v7045 => 5 1v7045 2v7515
2ndidx 6 1v3676 => 6 1v3676 2v4006
2ndidx 7 1v3977 => 7 1v3977 2v1707
2ndidx 8 1v7198 => 8 1v7198 2v7348
2ndidx 9 1v5879 => 9 1v5879 2v4949
2ndidx 10 1v52310 => 10 1v52310 2v95410
2ndidx 11 1v43311 => 11 1v43311 2v82011
2ndidx 12 1v28312 => 12 1v28312 2v83712
2ndidx 14 1v54514 => 14 1v54514 2v58314
2ndidx 15 1v5215 => 15 1v5215 2v32315
2ndidx 16 1v61416 => 16 1v61416 2v67916
2ndidx 17 1v80517 => 17 1v80517 2v45117
2ndidx 18 1v11518 => 18 1v11518 2v26918
2ndidx 19 1v21819 => 19 1v21819 2v61719
2ndidx 20 1v87820 => 20 1v87820 2v34520
2ndidx 21 1v51221 => 21 1v51221 2v96921
2ndidx 22 1v40822 => 22 1v40822 2v29122
2ndidx 24 1v71024 => 24 1v71024 2v14224
2ndidx 25 1v68225 => 25 1v68225 2v93425
2ndidx 26 1v62126 => 26 1v62126 2v96526
2ndidx 27 1v57427 => 27 1v57427 2v20427
2ndidx 28 1v29828 => 28 1v29828 2v13428
2ndidx 29 1v98329 => 29 1v98329 2v44429
2ndidx 30 1v14430 => 30 1v14430 2v15230
2ndidx 31 1v18731 => 31 1v18731 2v21531
2ndidx 32 1v832 => 32 1v832 2v69732
2ndidx 34 1v70134 => 34 1v70134 2v53734
2ndidx 35 1v41335 => 35 1v41335 2v6935
2ndidx 36 1v8636 => 36 1v8636 2v82236
2ndidx 37 1v67037 => 37 1v67037 2v37037
2ndidx 38 1v80638 => 38 1v80638 2v68838
2ndidx 39 1v2639 => 39 1v2639 2v6639
2ndidx 40 1v80240 => 40 1v80240 2v17140
2ndidx 41 1v55741 => 41 1v55741 2v84741
2ndidx 42 1v77742 => 42 1v77742 2v73042
2ndidx 44 1v64644 => 44 1v64644 2v49644
2ndidx 45 1v12045 => 45 1v12045 2v68445
2ndidx 46 1v37446 => 46 1v37446 2v6546
2ndidx 47 1v37047 => 47 1v37047 2v17447
2ndidx 48 1v82848 => 48 1v82848 2v86748
2ndidx 49 1v75949 => 49 1v75949 2v70349
2ndidx 50 1v84350 => 50 1v84350 2v94250
2ndidx 51 1v40151 => 51 1v40151 2v36251
2ndidx 52 1v36752 => 52 1v36752 2v30752
2ndidx 54 1v79954 => 54 1v79954 2v82954
2ndidx 55 1v53955 => 55 1v53955 2v37955
2ndidx 56 1v56056 => 56 1v56056 2v85856
2ndidx 57 1v57957 => 57 1v57957 2v2657
2ndidx 58 1v88758 => 58 1v88758 2v50758
2ndidx 59 1v34559 => 59 1v34559 2v89859
2ndidx 60 1v43060 => 60 1v43060 2v80160
2ndidx 61 1v84561 => 61 1v84561 2v32561
2ndidx 62 1v62462 => 62 1v62462 2v48062
2ndidx 64 1v5364 => 64 1v5364 2v73664
2ndidx 65 1v80965 => 65 1v80965 2v27065
2ndidx 66 1v17566 => 66 1v17566 2v83966
2ndidx 67 1v6167 => 67 1v6167 2v22267
2ndidx 68 1v35868 => 68 1v35868 2v50568
2ndidx 69 1v62769 => 69 1v62769 2v90669
2ndidx 70 1v4670 => 70 1v4670 2v98170
2ndidx 71 1v11971 => 71 1v11971 2v471
2ndidx 72 1v1472 => 72 1v1472 2v48072
2ndidx 74 1v87574 => 74 1v87574 2v63974
2ndidx 75 1v82775 => 75 1v82775 2v34475
2ndidx 76 1v59876 => 76 1v59876 2v56376
2ndidx 77 1v77077 => 77 1v77077 2v51677
2ndidx 78 1v53878 => 78 1v53878 2v54878
2ndidx 79 1v35779 => 79 1v35779 2v32279
2ndidx 80 1v3680 => 80 1v3680 2v37080
2ndidx 81 1v33181 => 81 1v33181 2v81581
2ndidx 82 1v76982 => 82 1v76982 2v66882
2ndidx 84 1v69684 => 84 1v69684 2v45284
2ndidx 85 1v40685 => 85 1v40685 2v185
2ndidx 86 1v39586 => 86 1v39586 2v32486
2ndidx 87 1v3687 => 87 1v3687 2v73887
2ndidx 88 1v16088 => 88 1v16088 2v7988
2ndidx 89 1v75989 => 89 1v75989 2v65789
2ndidx 90 1v3190 => 90 1v3190 2v78390
2ndidx 91 1v65091 => 91 1v65091 2v82491
2ndidx 92 1v85292 => 92 1v85292 2v6892
2ndidx 94 1v81394 => 94 1v81394 2v77594
2ndidx 95 1v8795 => 95 1v8795 2v69695
2ndidx 96 1v19696 => 96 1v19696 2v38096
2ndidx 97 1v7997 => 97 1v7997 2v75197
2ndidx 98 1v32398 => 98 1v32398 2v21798
2ndidx 99 1v3799 => 99 1v3799 2v70199
2ndIDX NULL
2ndidxnull 3 2v4143
2ndidxnull 13 2v41513
2ndidxnull 23 2v95323
2ndidxnull 33 2v28033
2ndidxnull 43 2v11543
2ndidxnull 53 2v16753
2ndidxnull 63 2v67663
2ndidxnull 73 2v40573
2ndidxnull 83 2v28183
2ndidxnull 93 2v54693
